### PR TITLE
Integrate with `auth-proxy` and `forward-auth`

### PR DESCRIPTION
**Pull request by natalian98**
This PR adds e2e integration with Traefik ForwardAuth middleware and charmed apps requesting the proxy.

### CODE REVIEW

File: templates/oauth2-proxy.toml.j2 [hunk 1]
```diff
@@ -1,4 +1,6 @@
 http_address="0.0.0.0:4180"
+ssl_insecure_skip_verify ="true"
```
Comment (natalian98) [inline templates/oauth2-proxy.toml.j2:2]: This is required as oauth2 proxy doesn't trust the hydra cert due to lack of `certificate_transfer` integration, I'll address this in a separate PR.

File: CONTRIBUTING.md
Comment (natalian98) [inline CONTRIBUTING.md:1]: I added integration tests to check if requests go through traefik to oauth2 proxy and if it correctly processes them (200 response when an endpoint is allowed or 403 otherwise).
That covers the app -> forward-auth -> oauth2 proxy flow.

I didn't add e2e tests because we would have to deploy the identity platform and that would take much longer to run, instead I added instructions here. Let me know if you think the platform should be part of the tests.

File: src/integrations.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+import logging
+from dataclasses import dataclass, field
+from typing import Any, List, Type, TypeAlias, Union
+
+from charms.oauth2_proxy_k8s.v0.auth_proxy import AuthProxyProvider
+
+logger = logging.getLogger(__name__)
+
+JsonSerializable: TypeAlias = Union[dict[str, Any], list[Any], int, str, float, bool, Type[None]]
```
Comment (wood-push-melon) [inline src/integrations.py:12]: We can remove this if it's not used anywhere.

File: src/integrations.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+@dataclass(frozen=True, slots=True)
+class AuthProxyIntegrationData:
+    """Data source from the auth-proxy integration."""
+
+    app_names: List[str] = field(default_factory=lambda: [])
+    allowed_endpoints: List[str] = field(default_factory=lambda: [])
+    headers: List[str] = field(default_factory=lambda: [])
+    authenticated_emails: List[str] = field(default_factory=lambda: [])
+    authenticated_email_domains: List[str] = field(default_factory=lambda: [])
```
Comment (wood-push-melon) [inline src/integrations.py:23]: Just a nitpick (feel free to ignore): `default_factory=lambda: []` -> `default_factory=list`

File: templates/oauth2-proxy.toml.j2 (hunk 1 above)
Comment (wood-push-melon) [reply templates/oauth2-proxy.toml.j2:2]: Just a format nitpick: we can remove the whitespace before the `=`.

File: README.md [hunk 4]
```diff
@@ -30,39 +30,69 @@ juju deploy oauth2-proxy-k8s --channel edge --trust
...
 
 ```shell
-juju integrate oauth2-proxy-k8s:oauth hydra
+juju config traefik-public enable_experimental_forward_auth=True
+juju integrate oauth2-proxy-k8s traefik-public:experimental-forward-auth
 ```
 
+### Auth Proxy
+
+OAuth2 Proxy can be integrated with downstream charmed operators
```
Comment (wood-push-melon) [inline README.md:62]: nitpick:

I think it should use `upstream` instead of `downstream`.

File: src/charm.py [hunk 5]
```diff
@@ -246,6 +315,45 @@ def _on_oauth_info_changed(self, event: OAuthInfoChangedEvent) -> None:
...
+
+        self._handle_status_update_config(event)
+
+        logger.info("Auth-proxy config has changed. Forward-auth relation will be updated")
+        self.forward_auth.update_forward_auth_config(self._forward_auth_config)
+
+    @log_event_handler(logger)
+    def _remove_auth_proxy_configuration(self, event: AuthProxyConfigRemovedEvent) -> None:
+        """Remove the auth-proxy-related config for a given relation."""
+        if not self.unit.is_leader():
```
Comment (wood-push-melon) [inline src/charm.py:351]: Just a question:

Why do we only allow the leader unit here?

Comment (natalian98) [reply src/charm.py:353]: Good catch, only the leader should write to the relation data but all units must reconfigure. Fixed

File: README.md (hunk 4 above)
Comment (natalian98) [reply README.md:62]: I'm referring to those as downstream operators since they receive a service from oauth2 proxy (upstream app). They consume it so are "lower" in the chain. I think the nomenclature here is correct.

File: src/charm.py [hunk 6]
```diff
@@ -246,6 +315,46 @@ def _on_oauth_info_changed(self, event: OAuthInfoChangedEvent) -> None:
...
+
+    @log_event_handler(logger)
+    def _remove_auth_proxy_configuration(self, event: AuthProxyConfigRemovedEvent) -> None:
+        """Remove the auth-proxy-related config for a given relation."""
+        self._handle_status_update_config(event)
+
+        if not self.unit.is_leader():
+            return
+
+        self.forward_auth.update_forward_auth_config(self._forward_auth_config)
```
Comment (nsklikas) [inline src/charm.py:356]: I think that the leadership check should be moved in the `update_forward_auth_config`

File: src/charm.py [hunk 7]
```diff
@@ -246,6 +315,46 @@ def _on_oauth_info_changed(self, event: OAuthInfoChangedEvent) -> None:
...
+    def _on_auth_proxy_config_changed(self, event: AuthProxyConfigChangedEvent) -> None:
+        if not self._oauth2_proxy_service_is_running:
+            self.unit.status = WaitingStatus("Waiting for OAuth2 Proxy service")
+            event.defer()
+            return
+
+        self._handle_status_update_config(event)
+
+        logger.info("Auth-proxy config has changed. Forward-auth relation will be updated")
+        self.forward_auth.update_forward_auth_config(self._forward_auth_config)
```
Comment (nsklikas) [inline src/charm.py:346]: Shouldn't we validate that this is the leader?

File: tests/integration/test_charm.py
Comment (nsklikas) [inline tests/integration/test_charm.py:1]: I think that we should have a test for scaling up/down `oauth2-proxy`

File: src/charm.py (hunk 6 above)
Comment (natalian98) [reply src/charm.py:352]: I just realized that the leadership is already checked in `forward-auth`:
https://github.com/canonical/oauth2-proxy-k8s-operator/blob/d97c61afc4eb350ca804453ba17bbb8ad1632c60/lib/charms/oauth2_proxy_k8s/v0/forward_auth.py#L566-L567

File: src/charm.py (hunk 7 above)
Comment (natalian98) [reply src/charm.py:346]: Leadership is already checked in `forward-auth`, see comment above

**Comment by natalian98** [comment]
> I understand that these libs where copied from oathkeeper, but they are using jsonchemas instead of pydantic. Maybe we should migrate them?

Yes, I opened https://github.com/canonical/oauth2-proxy-k8s-operator/issues/15 for that. I think that should be addressed together with any changes in the libs that may be required due to istio, so it all comes together in v1

**Review by wood-push-melon — APPROVED** [review]

File: tests/integration/test_charm.py [hunk 8]
```diff
@@ -102,3 +103,116 @@ async def test_health_checks(ops_test: OpsTest, lightkube_client: Client) -> Non
...
+    """Ensure that oauth2-proxy is able to provide forward-auth relation."""
+    await ops_test.model.applications[TRAEFIK].set_config({
+        "enable_experimental_forward_auth": "True"
+    })
+    await ops_test.model.integrate(f"{TRAEFIK}:experimental-forward-auth", APP_NAME)
+
+    await ops_test.model.wait_for_idle([TRAEFIK, APP_NAME], status="active", timeout=1000)
+
+
+@retry(
```
Comment (nsklikas) [inline tests/integration/test_charm.py:144]: question: why do we need the retry here?

Comment (natalian98) [reply tests/integration/test_charm.py:144]: for an edge case in which the config is not yet populated, this could happen if the oauth2 proxy reconfigures when a `relation-changed` is emitted

**Review by nsklikas — APPROVED** [review]
LGTM

@natalian98 do you think we should investigate the edge case you mentioned?

**Comment by natalian98** [comment]
> do you think we should investigate the edge case you mentioned?

Sure, I'm keeping the retry to reduce the chances of test flakiness, I created an issue to further evaluate that.
